Add Azure Key Provider for Encryption at Rest

ndragazis opened the pull request:
This PR introduces a new Key Provider to support Azure Key Vault as a Key Management System (KMS) for Encryption at Rest. The core design principle is the same as in the AWS and GCP key providers - an externally provided Vault key that is used to protect local data encryption keys (a process known as "key wrapping").

In more detail, this patch series consists of:
* Multiple Azure credential sources, offering a variety of authentication options (Service Principals, Managed Identities, environment variables, Azure CLI).
* The Azure host - the Key Vault endpoint bridge.
* The Azure Key Provider - the interface for the Azure host.
* Unit tests using real Azure resources (credentials and Vault keys).
* Log filtering logic to not expose sensitive data in the logs (plaintext keys, credentials, access tokens). 

This is part of the overall effort to support Azure deployments.

Testing done:
* Unit tests.
* Manual test on an Azure VM with a Managed Identity.
* Manual test with credentials from Azure CLI.
* Manual test of `--azure-hosts` cmdline option.
* Manual test of log filtering.

Remaining items:
- [x] Create necessary Azure resources for CI.
- [x] Merge pipeline changes (https://github.com/scylladb/scylla-pkg/pull/5201).

Closes https://github.com/scylladb/scylla-enterprise/issues/1077.

New feature. No backport is needed.

elcallio reviewed (changes requested):
Lots of code. But very neatly organised. Nice.
A few mostly stylistic remarks. Some of which can maybe be left for later.

ndragazis commented on utils/azure/identity/service_principal_credentials.cc:
```diff
@@ -23,8 +92,93 @@ service_principal_credentials::service_principal_credentials(const sstring& tena
+    http::experimental::client http_client{std::move(factory)};
+
+    sstring resp;
+    co_await http_client.make_request(std::move(req), [&](const http::reply& rep, input_stream<char>&& in) -> future<> {
```
The problem is that the `encryption::httpclient` lives under `encryption/`, but the Azure credentials are a more general utility living under `utils/`. I sent you an email the other day, proposing to move the client under `utils/` so that I can reuse it, but you raised concerns about not being general enough in its current form.

ndragazis commented:
v2 Changelog:
* Addressed review comments from v1. For the http client, I rebased on latest master and used the REST client, which is the encryption client moved into `utils/`. For the logging helpers in credentials I moved them into a separate class and I added a formatter for the credentials.
* Added a mock server for Azure Key Vault. All unit tests can now run locally, and some can be parameterized to also run with real resources.
* Added a few more local-only tests for IMDS, network errors, and error handling.
* Added a few fast retries for transient errors in IMDS, Entra, and Key Vault.
* Added the Azure Key Provider in the EaR ops guide.

ndragazis commented:
> I still don't like the member log methods

Even after separating them into a distinct logger class?

ndragazis commented:
v3 Changelog:
* Changed name prefix in test functions (`s/_/do_`).
* Added `ENABLE_AZURE_TEST` env var in tests to guard all Azure tests. Renamed `ENABLE_AZURE_REAL_TEST` into `ENABLE_AZURE_TEST_REAL`.

ndragazis commented:
> > I still don't like the member log methods
> 
> Even after separating them into a distinct logger class?

@elcallio do we want me to drop the logging helpers altogether?

elcallio commented:
I'd prefer the objects be able to format themselves, but not have logger methods, yes. Other than that small dissonance with my sensibilities, I wholly approve of this PR. :-)

ndragazis commented:
v4 Changelog:
* Removed logging helpers. Each credential type now explicitly prefixes its log messages with a formatted version of itself.
* Adjusted the log level in the `azure_host::impl` constructor from INFO to DEBUG.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17627/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17627/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17627/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/encryption_at_rest_test
#### Failed Tests (42/42):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_host.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_host.2.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_network_error.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_network_error.2.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_provider_with_both_secret_and_cert.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_provider_with_both_secret_and_cert.2.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_provider_with_certificate.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_provider_with_certificate.2.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_provider_with_imds.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17627/testReport/junit/boost.encryption_at_rest_test.test_azure_provider_with_imds.2.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 1 hr 54 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17962/) | :white_check_mark: | sif |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17962/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17962/) | :white_check_mark: | spider4.cloudius-systems.com |

#### Build Details:
- Duration: 22 min

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/) | :white_check_mark: | thor |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/) | :x: | spider9.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/)<br> :small_blue_diamond: test/boost/encryption_at_rest_test.cc | :white_check_mark: | spider4.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17963/)<br> :small_blue_diamond: test/boost/encryption_at_rest_test.cc | :white_check_mark: | spider3.cloudius-systems.com |

#### Failed Tests (4/8151):
- [test_kmip_provider_broken_sstables_on_restart.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17963/testReport/junit/boost.encryption_at_rest_test/cc/dev___test_py___dev___test_kmip_provider_broken_sstables_on_restart_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kmip_provider_broken_sstables_on_restart.dev.1)
- [test_kmip_provider.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17963/testReport/junit/boost.encryption_at_rest_test/cc/dev___test_py___dev___test_kmip_provider_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kmip_provider.dev.1)
- [test_kmip_provider_broken_config_on_restart.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17963/testReport/junit/boost.encryption_at_rest_test/cc/dev___test_py___dev___test_kmip_provider_broken_config_on_restart_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kmip_provider_broken_config_on_restart.dev.1)
- [test_kmip_provider_multiple_hosts.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17963/testReport/junit/boost.encryption_at_rest_test/cc/dev___test_py___dev___test_kmip_provider_multiple_hosts_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kmip_provider_multiple_hosts.dev.1)

#### Build Details:
- Duration: 1 hr 23 min

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/) | :white_check_mark: | thor |
| dev | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/) | :white_check_mark: | spider6.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/) | :white_check_mark: | spider6.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/)<br> :small_blue_diamond: boost/encryption_at_rest_test | :white_check_mark: | spider6.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/)<br> :small_blue_diamond: boost/encryption_at_rest_test | :white_check_mark: | spider9.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/235/)<br> :small_blue_diamond: boost/encryption_at_rest_test | :white_check_mark: | spider7.cloudius-systems.com |

#### Build Details:
- Duration: 55 min

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/240/) | :white_check_mark: | thor |
| dev | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/240/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/240/) | :x: | spider7.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/240/) | :white_check_mark: | spider6.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/240/) | :x: | spider6.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/240/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/240/) | :x: | spider3.cloudius-systems.com |

#### Failed Tests (9/351):
- [test_azure_provider_with_certificate_real.debug.1](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/debug___test_py___debug___test_azure_provider_with_certificate_real_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.debug.1)
- [test_azure_provider_with_certificate_real.debug.2](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/debug___test_py___debug___test_azure_provider_with_certificate_real_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.debug.2)
- [test_azure_provider_with_certificate_real.debug.3](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/debug___test_py___debug___test_azure_provider_with_certificate_real_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.debug.3)
- [test_azure_provider_with_certificate_real.dev.1](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/dev___test_py___dev___test_azure_provider_with_certificate_real_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.dev.1)
- [test_azure_provider_with_certificate_real.dev.2](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/dev___test_py___dev___test_azure_provider_with_certificate_real_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.dev.2)
- [test_azure_provider_with_certificate_real.dev.3](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/dev___test_py___dev___test_azure_provider_with_certificate_real_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.dev.3)
- [test_azure_provider_with_certificate_real.release.1](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/release___test_py___release___test_azure_provider_with_certificate_real_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.release.1)
- [test_azure_provider_with_certificate_real.release.2](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/release___test_py___release___test_azure_provider_with_certificate_real_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.release.2)
- [test_azure_provider_with_certificate_real.release.3](https://jenkins.scylladb.com//job/releng-testing/job/scylla-ci-debug/240/testReport/junit/boost.encryption_at_rest_test/cc/release___test_py___release___test_azure_provider_with_certificate_real_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_azure_provider_with_certificate_real.release.3)

#### Build Details:
- Duration: 21 min

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/) | :white_check_mark: | i-0df413201020bd9c6 |
| dev | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/)<br> :small_blue_diamond: test/boost/encryption_at_rest_test.cc | :white_check_mark: | spider8.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/)<br> :small_blue_diamond: test/boost/encryption_at_rest_test.cc | :white_check_mark: | spider1.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci-debug/251/)<br> :small_blue_diamond: test/boost/encryption_at_rest_test.cc | :white_check_mark: | spider4.cloudius-systems.com |

#### Build Details:
- Duration: 5 hr 36 min

ndragazis commented:
@pehala it will land on 2025.4. The milestone needs to be updated.

mykaul commented:
> @pehala it will land on 2025.4. The milestone needs to be updated.

CC @dani-tweig
